components/layout_2020/flow/root.rs

```diff
@@ -229,15 +200,165 @@ impl FragmentTreeRoot {
         ))
     }
 
-    pub fn bounding_box_of_border_boxes(&self) -> Rect<Au> {
-        let origin = Point2D::new(
-            Au::from_f32_px(self.bounding_box_of_border_boxes.origin.x.px()),
-            Au::from_f32_px(self.bounding_box_of_border_boxes.origin.y.px()),
-        );
-        let size = Size2D::new(
-            Au::from_f32_px(self.bounding_box_of_border_boxes.size.width.px()),
-            Au::from_f32_px(self.bounding_box_of_border_boxes.size.height.px()),
-        );
-        Rect::new(origin, size)
+    fn iterate_through_fragments(&self, iterator: &mut dyn FragmentIterator) {
+        fn do_iteration(
+            fragment: &Fragment,
+            containing_block: &PhysicalRect<Length>,
+            iterator: &mut dyn FragmentIterator,
+        ) -> bool {
+            if !iterator.process(fragment, containing_block) {
+                return false;
+            }
+
+            match fragment {
+                Fragment::Box(fragment) => {
+                    let new_containing_block = fragment
+                        .content_rect
+                        .to_physical(fragment.style.writing_mode, containing_block)
+                        .translate(containing_block.origin.to_vector());
+                    for child in &fragment.children {
+                        do_iteration(child, &new_containing_block, iterator);
+                    }
+                },
+                Fragment::Anonymous(fragment) => {
+                    let new_containing_block = fragment
+                        .rect
+                        .to_physical(fragment.mode, containing_block)
+                        .translate(containing_block.origin.to_vector());
+                    for child in &fragment.children {
+                        do_iteration(child, &new_containing_block, iterator);
+                    }
+                },
+                _ => {},
+            }
+
+            true
+        }
+
+        for child in &self.children {
+            if !do_iteration(child, &self.containing_block, iterator) {
+                break;
+            }
+        }
+    }
+
+    pub fn get_content_box_for_node(&self, requested_node: OpaqueNode) -> Rect<Au> {
+        let mut iterator = FragmentContentBoxIterator::new(requested_node);
+        self.iterate_through_fragments(&mut iterator);
+        let rect = &iterator.rect;
+
+        Rect::new(
+            Point2D::new(
+                Au::from_f32_px(rect.origin.x.px()),
+                Au::from_f32_px(rect.origin.y.px()),
+            ),
+            Size2D::new(
+                Au::from_f32_px(rect.size.width.px()),
+                Au::from_f32_px(rect.size.height.px()),
```

Comment:
It seems silly to do a lossy conversion from floating point to integer here when all consumers end up doing the opposite one. For example:

https://github.com/servo/servo/blob/5f55cd5d71df9c555fbc24777168396ddd539f28/components/script/dom/element.rs#L2131-L2138

We can probably avoid it by changing `Rect<Au>` types to `Rect<Length>` in the query response, and move the conversion to Layout 2013 when creating that response. But let’s keep this for a future PR.

Reply:
Okay. These APIs are supposed to return rounded integers, but it's true that the conversion could be done in the script task.

Reply:
Okay. I'm happy to do this in a followup.